db/db_impl_write.cc

```diff
@@ -519,30 +629,52 @@ Status DBImpl::WriteToWAL(const WriteThread::WriteGroup& write_group,
     // and the batch is not wanting to be truncated
     merged_batch = leader->batch;
     leader->log_used = logfile_number_;
-    write_with_wal = 1;
+    *write_with_wal = 1;
   } else {
     // WAL needs all of the batches flattened into a single batch.
     // We could avoid copying here with an iov-like AddRecord
     // interface
-    merged_batch = &tmp_batch_;
+    merged_batch = tmp_batch;
     for (auto writer : write_group) {
       if (writer->ShouldWriteToWAL()) {
         WriteBatchInternal::Append(merged_batch, writer->batch,
                                    /*WAL_only*/ true);
-        write_with_wal++;
+        (*write_with_wal)++;
       }
       writer->log_used = logfile_number_;
     }
   }
+  return merged_batch;
+}
 
-  WriteBatchInternal::SetSequence(merged_batch, sequence);
-
-  Slice log_entry = WriteBatchInternal::Contents(merged_batch);
-  status = log_writer->AddRecord(log_entry);
+Status DBImpl::WriteToWAL(const WriteBatch &merged_batch,
+                          uint64_t* log_used, uint64_t* log_size) {
+  assert(log_size != nullptr);
+  Slice log_entry = WriteBatchInternal::Contents(&merged_batch);
+  *log_size = log_entry.size();
+  log::Writer* log_writer = logs_.back().writer;
```

Comment:
For the normal path, is it inside DB mutex? I moved it from here to inside the DB mutex to avoid a data race bug: https://github.com/facebook/rocksdb/commit/5dad9d6d28cef6a3df1070a6f2f09de170b22bf2

Are we getting the bug back here?